### PR TITLE
Convert to boost::beast

### PR DESCRIPTION
**Pull request by seelabs**
After this patch is accepted rippled will require boost 1.66 or later.

### CODE REVIEW

**Comment by seelabs** [comment]
Windows unit test asserts

**Comment by HowardHinnant** [comment]
Which assert?

**Comment by HowardHinnant** [comment]
Oh, inside boost/beast.  Interesting I'm not seeing that on macOS.

**Comment by HowardHinnant** [comment]
This is the line that is asserting:  https://github.com/boostorg/beast/blob/develop/include/boost/beast/websocket/impl/read.ipp#L638

Notes:

1.  This code already has fairly extensive changes since boost 1.66.
2.  This code path is not exercised during unit tests on macOS, but obviously is on Windows.

Questions:

1.  Is it exercised on Windows desktop or just by Appveyor?
2.  Why is it not exercised on macOS?

**Comment by seelabs** [comment]
@HowardHinnant I do exercise that line in linux. I'm currently trying to set up a windows VM that I can compile rippled on.

**Comment by HowardHinnant** [comment]
Interesting, thanks.  I'll try to figure out why it isn't exercised on macOS.

**Comment by HowardHinnant** [comment]
I am mistaken above:  This assert _is_ exercised on macOS, but does not fire.

**Comment by HowardHinnant** [comment]
This PR looks fine to me.  But I'm holding off on the approve until we learn more about the Windows unittest failure.  I've taken a cursory look, but have been unsuccessful at a diagnosis using only macOS.

**Comment by seelabs** [comment]
Pushed a fix for the window's assert (thanks you @miguelportilla!)

File: src/test/jtx/impl/WSClient.cpp [hunk 1]
```diff
@@ -117,7 +117,9 @@ class WSClientImpl : public WSClient
             error_code ec;
             if (!peerClosed_)
             {
-                ws_.async_close({}, [&](error_code ec) { stream_.close(ec); });
+                ws_.async_close({}, strand_.wrap([&](error_code ec) {
+                    stream_.close(ec);
```
Comment (miguelportilla) [inline src/test/jtx/impl/WSClient.cpp]: It is not necessary to call `stream_.close(ec)` as its already closed with `ws_async_close`. However, calling `stream_.cancel(ec)` is a good idea to stop all asynchronous operations associated with the socket.

Comment (seelabs) [reply src/test/jtx/impl/WSClient.cpp]: Fixed and pushed.

File: src/ripple/rpc/json_body.h [hunk 2]
```diff
@@ -57,19 +59,50 @@ struct json_body
...
+            ec.assign(0, ec.category());
+        }
+
+        boost::optional<std::pair<const_buffers_type, bool>>
+        get(boost::beast::error_code& ec)
         {
+            ec.assign(0, ec.category());
+            return {{const_buffers_type{
+                body_string_.data(), body_string_.size()}, false}};
         }
```
Comment (miguelportilla) [inline src/ripple/rpc/json_body.h]: Looks good. I wonder if storing the buffer might be a better option.

eg.
```
    class writer
    {
        boost::asio::const_buffer buffer_;

    public:
        using const_buffers_type =
            boost::asio::const_buffer;

        template<bool isRequest, class Fields>
        explicit
        writer(boost::beast::http::message<isRequest,
                json_body, Fields> const& msg)
        {
            auto s {to_string(msg.body())};
            buffer_ = const_buffers_type{s.data(), s.size()};
        }

        void
        init(boost::beast::error_code& ec)
        {
            ec.assign(0, ec.category());
        }

        boost::optional<std::pair<const_buffers_type, bool>>
        get(boost::beast::error_code& ec)
        {
            ec.assign(0, ec.category());
            return {{buffer_, false}};
        }
    };
```

Comment (seelabs) [reply src/ripple/rpc/json_body.h]: `const_buffer` doesn't own it's memory, so I don't think that works (it would point to the re-claimed string memory).

Comment (miguelportilla) [reply src/ripple/rpc/json_body.h]: You're correct, I forgot const_buffer doesn't own the memory.

**Comment by seelabs** [comment]
In 0.90.0-b5

**Pull request closed without merging**